### PR TITLE
Upgrade to drupal-code-generator v3

### PR DESCRIPTION
**Pull request by Chi-teck**
Drush currently uses DCG 2 to generate code.

### Remaining tasks
- [x] Update `drush generate` command to use PHP attributes instead of annotations
- [x] Update `drush:command-file` generator
- [x] Figure out a way to register third-party generators
- [x] Update Drush tests
- [ ] Implement completion for generator names

The big question is how to deliver this change. Here are a couple complications.
1. DCG 3 is not API compatible with DCG 2. 
2. DCG 3 requires PHP 8.1 and Drupal 10.

Basically there are two options.
1. Wait for a new major Drush release that would only support Drupal 10.
2. Make `drush generate` compatible with both DCG versions.

### CODE REVIEW

**Comment by Chi-teck** [comment]
This kind of critical, because Drupal 10 has been already released and code generated by DCG 2 is not fully compatible with it. In order to support multiple Drupal versions DCG would have to generate code with deprecations .

**Comment by weitzman** [comment]
I dont feel strongly about it, but I think 1. is the best way forward. We could branch for Drush 12 now, with a target release date just before Drupalcon Pittsburgh (early June). Drush 12 would be Drupal 10 minimum. Here is the todo list for Drush 12 https://github.com/drush-ops/drush/milestone/7

**Comment by weitzman** [comment]
_Update: ignore below. I had merged composer changes into 11.x by accident._

I changed the base of this PR to 12.x It says it has no merge conflicts but I dont see how composer.lock could not conflict. Please reroll if needed when you get a chance.

**Comment by Chi-teck** [comment]
> Figure out a way to register third-party generators

Modules can subscribe to generator info event and register their generators. I am not sure if this is enough to resolve this task.

File: src/Commands/generate/GenerateCommands.php [hunk 1]
```diff
@@ -20,40 +25,42 @@ class GenerateCommands extends DrushCommands implements AutoloaderAwareInterface
...
     {
+        // @todo Figure out a way to inject the container.
+        $container = \Drupal::getContainer();
+
+        // @todo Implement discovery for third-party generators.
+        $application = Application::create($container);
+        $application->setAutoExit(FALSE);
+
+        $application->addCommands([
+            new MigrationGenerator(),
```
Comment (weitzman) [inline src/Commands/generate/GenerateCommands.php]: Any interest in moving Migration generator in DCG? If not thats , OK.

**Review by weitzman — COMMENTED** [review]
How does a global generator subscribe to the event? Here is what we say about authoring global generators today https://www.drush.org/11.x/generators/#global-generators

File: src/Commands/generate/GenerateCommands.php (hunk 1 above)
Comment (Chi-teck) [reply src/Commands/generate/GenerateCommands.php]: DCG already have a generator for MigrationSource plugins. We could add a simple migration generator under Yaml namespace.

**Comment by Chi-teck** [comment]
> How does a global generator subscribe to the event? Here is what we say about authoring global generators today https://www.drush.org/11.x/generators/#global-generators

Those need to be registered in the generate command before `$application->run();`

File: src/Commands/generate/GenerateCommands.php (hunk 1 above)
Comment (Chi-teck) [reply src/Commands/generate/GenerateCommands.php]: https://github.com/Chi-teck/drupal-code-generator/commit/6347e6ffea446194e37a8fb2db9067d29b0bd676

**Comment by weitzman** [comment]
I updated the dcf generator a bit in https://github.com/drush-ops/drush/pull/5442. Still needs port to DCG3.

**Comment by weitzman** [comment]
For global generators, is there any reason we can't keep using the code in `discoverPsr4Generators` - https://github.com/drush-ops/drush/pull/5371/files#diff-14a8581695da2d03094b84103528a480af8e2891e0590de222aa9b04a8d4d7f2L137-L144. That can be called right from the command method.

**Comment by Chi-teck** [comment]
> I updated the dcf generator a bit in #5442. Still needs port to DCG3.

The dcf command has been updated to DCG3. Needs review.

**Comment by Chi-teck** [comment]
> For global generators, is there any reason we can't keep using the code in `discoverPsr4Generators` - https://github.com/drush-ops/drush/pull/5371/files#diff-14a8581695da2d03094b84103528a480af8e2891e0590de222aa9b04a8d4d7f2L137-L144. That can be called right from the command method.

The link seems broken. Anyway we can use any discovery method as long as it can ignore DCG 2 generators. I would prefer something less magic, something that does not rely on reflection API and loading generator classes. That would allow to avoid difficulties in future major updates of DCG and symfony/console.

**Comment by weitzman** [comment]
The diff wont show because it is inside the file thats deleted in the PR.

IMO we should merge this without major changes to generator discovery in drush (both global and module). Then we work on those in #4685

**Comment by Chi-teck** [comment]
I've removed `discoverDrushGenerators`. I don't think it's worth implementing a discovery method for just a couple of generators. Also replaced `adjustDCGGenerators` with event subscriber.
